core/opendaq/reader/src/block_reader_impl.cpp

```diff
@@ -33,14 +46,18 @@ BlockReaderImpl::BlockReaderImpl(const ReaderConfigPtr& readerConfig,
 
 BlockReaderImpl::BlockReaderImpl(BlockReaderImpl* old,
                                  SampleType valueReadType,
-                                 SampleType domainReadType)
+                                 SampleType domainReadType,
+                                 SizeT blockSize)
     : Super(old, valueReadType, domainReadType)
-    , blockSize(old->blockSize)
+    , blockSize(blockSize)
     , info(old->info)
 {
-    this->internalAddRef();
-
-    readDescriptorFromPort();
+    // if on descreption change callback was set
+    // readDescriptorFromPort method have to be called from this callback
+    packets = old->packets;
+    availableSamples = old->availableSamples;
+    if (!changeCallback.assigned())
+        readDescriptorFromPort();
```

Comment:
As far as I can tell this method sets up a listener and reads the latest descriptor. Shouldn't this always be called?

Reply:
listener sets from parent class 
```
explicit ReaderImpl(ReaderImpl* old,
                        SampleType valueReadType,
                        SampleType domainReadType))
```
                        
if reader created from signal,  our packet notification is `SameThread. If we are configuring callback on description changed, where we are recreating reader from existing. firstly signal send packet (here signal has lock). Then we are creating new reader, where we are trying to get signal description in readDescriptorFromPort which also has lock. This situation makes makes a deadlock. To escape this problem we are not calling readDescriptorFromPort  but setting reader types from old reader